src/arch/riscv/isa.cc

```diff
@@ -995,6 +1272,93 @@ ISA::getFaultHandlerAddr(RegIndex idx, uint64_t cause, bool intr) const
     return addr;
 }
 
+// V-bit utilities (H-extension)
+bool virtualizationEnabled(ExecContext *xc) {
+    return xc->readMiscReg(MISCREG_VIRT);
+}
+bool virtualizationEnabled(ThreadContext *tc) {
+    return tc->readMiscReg(MISCREG_VIRT);
+}
+
+void setV(ExecContext *xc) {
+    assert(!virtualizationEnabled(xc));
+    xc->setMiscReg(MISCREG_VIRT, 1);
+}
+void setV(ThreadContext *tc) {
+    assert(!virtualizationEnabled(tc));
+    tc->setMiscReg(MISCREG_VIRT, 1);
+}
+
+void resetV(ExecContext *xc) {
+    assert(virtualizationEnabled(xc));
+    xc->setMiscReg(MISCREG_VIRT, 0);
+}
+void resetV(ThreadContext *tc) {
+    assert(virtualizationEnabled(tc));
+    tc->setMiscReg(MISCREG_VIRT, 0);
+}
+
+// FPU status update function
+Fault updateFPUStatus(ExecContext *xc, ExtMachInst machInst, bool set_dirty) {
+
+
+    MISA misa = xc->readMiscReg(MISCREG_ISA);
+    STATUS status = xc->readMiscReg(MISCREG_STATUS);
+    STATUS vsstatus = misa.rvh && virtualizationEnabled(xc) ?
+        xc->readMiscReg(MISCREG_VSSTATUS) : 0;
+
+    if (status.fs == FPUStatus::OFF ||
+        (misa.rvh && virtualizationEnabled(xc) &&
+         vsstatus.fs == FPUStatus::OFF))
+        return std::make_shared<IllegalInstFault>("FPU is off", machInst);
+
+    if (set_dirty) {
+        status.fs = FPUStatus::DIRTY;
+        xc->setMiscReg(MISCREG_STATUS, status);
+
+        if (misa.rvh && virtualizationEnabled(xc)) {
+            vsstatus.fs = FPUStatus::DIRTY;
+            xc->setMiscReg(MISCREG_VSSTATUS, vsstatus);
+        }
+    }
+
+    return NoFault;
+}
+
+// VPU status update function
+Fault updateVPUStatus(
+    ExecContext *xc, ExtMachInst machInst, bool set_dirty, bool check_vill) {
+
+    MISA misa = xc->readMiscReg(MISCREG_ISA);
+    STATUS status = xc->readMiscReg(MISCREG_STATUS);
+    STATUS vsstatus = misa.rvh && virtualizationEnabled(xc) ?
+        xc->readMiscReg(MISCREG_VSSTATUS) : 0;
+
+    if (!misa.rvv || status.vs == VPUStatus::OFF ||
+        (misa.rvh && virtualizationEnabled(xc) &&
+         vsstatus.vs == VPUStatus::OFF))
+        return std::make_shared<IllegalInstFault>(
+            "RVV is disabled or VPU is off", machInst);
```

Comment:
Must have curly braces since the inner block is multiple lines. However, IMO all `if` statements should always have curly braces. Again, I have spent many hours tracking down bugs because someone tried to save 2 characters. It ends up being not worth it in the end :)